hackathon_site/review/admin.py

```diff
@@ -239,14 +256,90 @@ def has_delete_permission(self, request, obj=None):
+        ascending, so teams with all applications submitted earlier get reviewed
+        first.
+
+        Each user can be assigned as a reviewer to one team. When that user
+        requests a new team, they are unassigned from the previous one. The assignment
+        will expire after 20 minutes.
         """
-        return HttpResponse(f"Hello there, {request.user.first_name}")
 
-    def get_urls(self):
-        from django.urls import path
+        if not self.has_change_permission(request):
+            raise PermissionDenied
+
+        active_reviewers_set_cache_key = "admin:assign_to_team:active_reviewers"
+
+        def build_reviewer_assignment_cache_key(user_id):
+            """
+            Build a key that stores which team a reviewer is assigned to
+            """
+            return f"admin:assign_to_team:user-{user_id}"
+
+        with transaction.atomic():
+            # Set of reviewers IDs that are currently reviewing
+            active_reviewers = cache.get(active_reviewers_set_cache_key, set())
+
+            # Get a dictionary mapping user id -> assigned team
+            assigned_teams = cache.get_many(
+                [
+                    build_reviewer_assignment_cache_key(user_id)
+                    for user_id in active_reviewers
+                ]
+            )
+
+            current_user_cache_key = build_reviewer_assignment_cache_key(
+                request.user.id
+            )
+
+            queryset = (
+                self.get_queryset(request)
+                .filter(applications__isnull=False)  # Exclude any teams that are empty
+                # Get any team where at least one application is missing a review
+                .filter(applications__review__isnull=True)
+                .exclude(
+                    id__in=list(assigned_teams.values())
+                )  # Exclude teams that are currently assigned
+                .order_by(
+                    "most_recent_submission"
+                )  # Ascending order, annotated in self.get_queryset
+            )
+
+            team = queryset.first()
+
+            if team is None:
+                # No more teams left to review
+                cache.delete(current_user_cache_key)
+
+                messages.add_message(
+                    request,
+                    messages.INFO,
+                    "No more teams remaining. Note that some reviews may still be in progress, "
+                    "so check back later to ensure all were completed",
+                )
+
+                return HttpResponseRedirect(
+                    reverse("admin:review_teamreview_changelist")
+                )
 
+            # Assign the user to their team in the cache, 20 minute TTL
+            cache.set(current_user_cache_key, team.id, timeout=60 * 20)
+
+            # Add the user to the set of active reviewers
+            active_reviewers.add(request.user.id)
```

Comment:
Shouldn't you check if the user is already in the list of `active_reviewers`?

Reply:
It's a set, so doesn't matter